packages/datadog-plugin-dns/src/index.js

```diff
@@ -1,201 +1,88 @@
-}
 
-function createWrapLookupService (tracer, config) {
-  return function wrapLookupService (lookupService) {
-    return function lookupServiceWithTrace (address, port, callback) {
-      if (!isArgsValid(arguments, 3)) return lookupService.apply(this, arguments)
-
-      const span = startSpan(tracer, config, 'dns.lookup_service', {
-        'resource.name': `${address}:${port}`,
-        'dns.address': address,
-        'dns.port': port
-      })
-
-      wrapArgs(span, arguments)
-
-      return tracer.scope().activate(span, () => lookupService.apply(this, arguments))
-    }
+  addSubs (func, start, asyncEnd = defaultAsyncEnd) {
+    this.addSub(`apm:dns:${func}:start`, start)
+    this.addSub(`apm:dns:${func}:end`, this.exit.bind(this))
+    this.addSub(`apm:dns:${func}:error`, errorHandler)
+    this.addSub(`apm:dns:${func}:async-end`, asyncEnd)
   }
-}
 
-function createWrapResolve (tracer, config) {
-  return function wrapResolve (resolve) {
-    return function resolveWithTrace (hostname, rrtype, callback) {
-      if (!isArgsValid(arguments, 2)) return resolve.apply(this, arguments)
-
-      if (typeof rrtype !== 'string') {
-        rrtype = 'A'
-      }
-
-      const span = wrapResolver(tracer, config, rrtype, arguments)
-
-      return tracer.scope().activate(span, () => resolve.apply(this, arguments))
-    }
-  }
-}
-
-function createWrapResolver (tracer, config, rrtype) {
-  return function wrapResolve (resolve) {
-    return function resolveWithTrace (hostname, callback) {
-      if (!isArgsValid(arguments, 2)) return resolve.apply(this, arguments)
-
-      const span = wrapResolver(tracer, config, rrtype, arguments)
-
-      return tracer.scope().activate(span, () => resolve.apply(this, arguments))
+  startSpan (name, customTags, store) {
+    const tags = {
+      'service.name': this.config.service || this.tracer._service,
+      'span.kind': 'client'
     }
-  }
-}
-
-function createWrapReverse (tracer, config) {
-  return function wrapReverse (reverse) {
-    return function reverseWithTrace (ip, callback) {
-      if (!isArgsValid(arguments, 2)) return reverse.apply(this, arguments)
-
-      const span = startSpan(tracer, config, 'dns.reverse', {
-        'resource.name': ip,
-        'dns.ip': ip
-      })
-
-      wrapArgs(span, arguments)
-
-      return tracer.scope().activate(span, () => reverse.apply(this, arguments))
+    for (const tag in customTags) {
+      tags[tag] = customTags[tag]
     }
+    return this.tracer.startSpan(name, {
+      childOf: store ? store.span : null,
+      tags
+    })
   }
-}
-
-function startSpan (tracer, config, operation, tags) {
-  const childOf = tracer.scope().active()
-  const span = tracer.startSpan(operation, {
-    childOf,
-    tags: Object.assign({
-      'service.name': config.service || tracer._service,
-      'span.kind': 'client'
-    }, tags)
-  })
-
-  analyticsSampler.sample(span, config.measured)
 
-  return span
-}
+  constructor (...args) {
+    super(...args)
```

Comment:
Calling a method on the subclass from the base class instead would avoid having to repeat this in every plugin.